ml-proto/host/encode.ml

```diff
@@ -0,0 +1,401 @@
+let put_string s bs = Buffer.add_string s.buf bs
+let patch s pos b = s.patches := (pos, b) :: !(s.patches)
+
+let to_string s =
+  let bs = Buffer.to_bytes s.buf in
+  List.iter (fun (pos, b) -> Bytes.set bs pos b) !(s.patches);
+  Bytes.to_string bs
+
+
+(* Encoding *)
+
+let encode m =
+  let s = stream () in
+
+  let module E = struct
+    (* Generic values *)
+
+    let u8 i = put s (Char.chr (i land 0xff))
+    let u16 i = u8 (i land 0xff); u8 (i lsr 8)
+    let u32 i =
+      Int32.(u16 (to_int (logand i 0xffffl));
+             u16 (to_int (shift_right i 16)))
+    let u64 i =
+      Int64.(u32 (to_int32 (logand i 0xffffffffL));
+             u32 (to_int32 (shift_right i 32)))
+
+    let rec vu64 i =
+      let b = Int64.(to_int (logand i 0x7fL)) in
+      if i < 128L then u8 b
+      else (u8 (b lor 0x80); vu64 (Int64.shift_right i 7))
+
+    let rec vs64 i =
+      let b = Int64.(to_int (logand i 0x7fL)) in
+      if -64L <= i && i < 64L then u8 b
+      else (u8 (b lor 0x80); vs64 (Int64.shift_right i 7))
+
+    let vu32 i = vu64 (Int64.of_int32 i)
+    let vs32 i = vs64 (Int64.of_int32 i)
+    let vu i = vu64 (Int64.of_int i)
+    let f32 x = u32 (F32.to_bits x)
+    let f64 x = u64 (F64.to_bits x)
+
+    let bool b = u8 (if b then 1 else 0)
+    let string bs = vu (String.length bs); put_string s bs
+    let list f xs = List.iter f xs
+    let opt f xo = Lib.Option.app f xo
+    let vec f xs = vu (List.length xs); list f xs
+    let vec1 f xo = bool (xo <> None); opt f xo
+
+    let gap () = let p = pos s in u32 0l; p
+    let patch_gap p n =
+      assert (n <= 0x0fff_ffff); (* Strings cannot excess 2G anyway *)
+      let lsb i = Char.chr (i land 0xff) in
+      patch s p (lsb (n lor 0x80));
+      patch s (p + 1) (lsb ((n lsr 7) lor 0x80));
+      patch s (p + 2) (lsb ((n lsr 14) lor 0x80));
+      patch s (p + 3) (lsb (n lsr 21))
+
+    (* Types *)
+
+    open Types
+
+    let value_type = function
+      | Int32Type -> u8 0x01
+      | Int64Type -> u8 0x02
+      | Float32Type -> u8 0x03
+      | Float64Type -> u8 0x04
+
+    let expr_type t = vec1 value_type t
+
+    let func_type = function
+      | {ins; out} -> u8 0x05; vec value_type ins; expr_type out
+
+    (* Expressions *)
+
+    open Source
+    open Kernel
+    open Ast
+
+    let op n = u8 n
+    let arity xs = vu (List.length xs)
+    let arity1 xo = bool (xo <> None)
+
+    let memop off align = vu align; vu64 off  (*TODO: to be resolved*)
+
+    let var x = vu x.it
+    let var32 x = vu32 (Int32.of_int x.it)
+
+    let rec expr e =
+      match e.it with
+      | Nop -> op 0x00
+      | Block es -> op 0x01; list expr es; op 0x17
+      | Loop es -> op 0x02; list expr es; op 0x17
+      | If (e, es1, es2) ->
+        expr e; op 0x03; list expr es1;
+        if es2 <> [] then op 0x04; list expr es2; op 0x17
+      | Select (e1, e2, e3) -> expr e1; expr e2; expr e3; op 0x05
+      | Br (x, eo) -> opt expr eo; op 0x06; arity1 eo; var x
```

Comment:
The expression is not optional in the binary encoding, one expression is expected which may be a `nop`. The arity is a constant 1 and not necessary but include as a concession.